✨ Feat: 상품 검색 API, 조회 API 추가

hyunmuam opened the pull request:
<!-- 
체크하려면 괄호 안에 "x"를 입력하세요. 
각 규칙은 Convention 문서에 있습니다.
PR 제목에 쓰는 prefix는 다음과 같습니다.

✨ Feat : 새로운 기능
🔨️ Refactor : 코드 리팩토링
🐎 Perf : 성능을 향상
🐛 Fix : 버그를 고칠 때
🧪 Test : 테스트 코드
🚜 Rename : 파일 이름 변경 혹은 구조를 변경
🚀 Release : 배포 / 개발 작업과 관련된 모든 것
🔥 Remove : 코드 또는 파일 제거
📚 Docs : 문서
📝 Chore : 사소한 코드 또는 언어를 변경 기타 변경사항 (빌드 스크립트 수정, 패키지 매니징 설정 등)

-->

## PR 타입
<!-- 어떤 유형의 PR인지 체크해주세요. -->

<!-- 체크하려면 괄호 안에 "x"를 입력하세요. -->
- [x] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Performance Improvement
- [ ] Bugfix
- [ ] Test Code
- [ ] Code style update (formatting, local variables)
- [ ] Other... Please describe:



## 개요 
1. API명세서의 상품 조회 기능 추가
2. 구현된 코드를 테스트하는 테스트 코드 추가

## 작업 및 변경 사항 
- 상품 검색 및 정렬 기능 구현
- 상품 상세 조회 기능 구현
- Repository계층 단위 테스트 코드 추가
- Service계층 단위 테스트 코드 추가
- Controller계층 통합 테스트 코드 추가

closes #3

hyunmuam committed:
[feat] 상품 조회 기능 추가

1. 상품 검색 기능 추가
2. 상품 상세 조회 기능 추가

Fixes: #3

hyunmuam committed:
[refactor] 목록 조회시 반환할 Dto 추가

Entity를 List의 형태 그대로 반환하는 것이아닌, 목록조회 API 명세서에 나온 것과 같이 ProductSimple라는 Dto와 이 Dto를 담을 ListDto를 추가.

Related to: #3

hyunmuam committed:
[test] 상품 조회기능 테스트 코드 추가.

Repository, Service계층에 단위테스트 코드 추가, Controller계층 은 통합테스트로하여 상품 주문조회 기능에 대한 테스트 완료

Resolves: #3

m1naworld commented on src/main/java/com/hanbang/e/product/dto/ProductDetails.java:
```diff
@@ -0,0 +1,49 @@
+import com.hanbang.e.product.entity.Product;
+
+import lombok.Builder;
+
+public class ProductDetails {
+	private Long productId;
+
+	private String brand;
+
+	private String name;
+
+	private Long price;
+
+	private String productimg;
+
+	private int stock;
+
+	private int sales;
+
+	private Boolean onSale;
+
```
저희 dto 부분 필드들은 한줄 띄지말고 붙힐까요?? 띄어져 있는 걸 보니 마음이.. 심란해요 🤣 변덕 죄송해요 🥹

hyunmuam replied:
@m1naworld 확인했습니다. 붙여도 될 것 같습니다. 저희 노션의 컨벤션에서도 이 항목을 추가하겠습니다.

m1naworld commented on src/main/java/com/hanbang/e/product/dto/ProductSimple.java:
```diff
@@ -0,0 +1,39 @@
+		this.productimg = productimg;
+	}
+
+	public static ProductSimple from(Product product) {
+		return ProductSimple.builder()
+			.productId(product.getProductId())
+			.brand(product.getBrand().getBrandName())
+			.name(product.getProductName())
+			.price(product.getPrice())
+			.productimg(product.getImg())
+			.build();
+	}
```
이부분은  빌더와 정적 팩토리 메소드를 혼합한 걸로 보이는데, 이렇게 하신 이유가 있을까욥?!

hyunmuam replied:
@m1naworld 음... 일단 이렇게 코드를 짠이유에 대해 말씀드리자면
1. Dto -> Entity의 컨벤션은 존재했으나 Entity -> Dto의 과정에 대한 컨벤션이 존재하지 않았습니다.
2. `from`이라는 메소드는 매개변수를 1개를 받고 `ProductSimple`을 반환해주기 때문에 정적 팩토리 메서드를 사용했습니다.
3. `from`메소드의 역할은 말그대로 `ProductSimple`을 생성하고 반환해주는 역할이어서, 일반 생성자로 만들려했으나 생성자에 `@Builder` 어노테이션이 붙어있어서 빌더패턴으로 만들게 되었습니다. 

만약 Dto에 `@Builder`를 사용하지 않았을 경우 다음과 같은 코드가 만들어 질 수 있을 것 같습니다.
```Java
public static ProductSimple from(Product product) {
    final Long productId = product.getProductID();
    final String brand = product.getBrand().getBrandName();
    final String name = product.getProductName();
    final Long price = product.getPrice();
    final String productimg = product.getProductimg();
    return new ProductSimple(productId, brand, name, price, productimg);
}
```
어떤점이 더 좋을지 한번 생각해봐야겠네요. 민아님의 의견은 어떠신가요?

m1naworld replied:
@tjvm0877 이유 설명 감사합니다! 두 코드를 함께보니 저는 현재 진행하신 방향이 가시성이 좋다는 의견입니다.

m1naworld commented on src/main/java/com/hanbang/e/product/repository/ProductRepository.java:
```diff
@@ -8,9 +8,9 @@
+	List<Product> findByProductNameContainingOrderByPriceAsc(String productName, Pageable pageable);
 
-	List<Product> findByProductNameContainingOrderBySalesDesc(String search, Pageable pageable);
+	List<Product> findByProductNameContainingOrderBySalesDesc(String productName, Pageable pageable);
```
더 직관적이네요!! 굿!!

m1naworld commented on src/main/java/com/hanbang/e/product/controller/ProductController.java:
```diff
@@ -25,15 +27,15 @@ public class ProductController {
-		return new ResponseEntity(response, HttpStatus.OK);
+		ProductDetails response = productService.getProductDetails(productId);
+		return new ResponseEntity(new ResponseDto("success", "상세 조회 성공", response), HttpStatus.OK);
 	}
```
컨트롤러 부분에서 반환하는 ResponseDto의 값이 보여서 너무 좋은데요?!

m1naworld reviewed (approved):
고생하셨습니다 👏🏻 조회API와 테스트 코드 대박이네욥! 컨트롤러부분 테스트 코드는 하나도 이해 안되는거 보니 얼른 강의를 들어야겠습니다 👀

hyunmuam committed:
[style] Dto관련 컨벤션 변경에 따른 코드 수정

- Dto클래스 이름 변경
- Dto의 필드간 빈줄 제거

Related to: #3

hyunmuam committed:
[rename] tableInit.sql 이름 변경

다른 도메인의 테스트에서도 AutoIncrement로 인해 테이블을 만들어주는 SQL문을 사용하여 충돌을 방지하고자 "도메인 이름+ Domain + TableInit"의 형태로 파일 이름 변경

Related to: #3

hyunmuam committed:
[style]  와일드카드(*)를 사용하는 import 변경

와일드 카드를 싱글 import로 변경

Related to: #3
